tests/test_base_device.py

```diff
@@ -0,0 +1,947 @@
+# Copyright 2021-2024 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This module contains tests for the base Qiskit device for the new PennyLane device API
+"""
+
+import numpy as np
+import pytest
+import inspect
+from unittest.mock import patch, Mock
+
+import pennylane as qml
+from pennylane.tape.qscript import QuantumScript
+
+from pennylane_qiskit import AerDevice
+from pennylane_qiskit.qiskit_device2 import (
+    QiskitDevice2,
+    qiskit_session,
+    accepted_sample_measurement,
+    split_measurement_types,
+    qiskit_options_to_flat_dict,
+)
+from pennylane_qiskit.converter import circuit_to_qiskit, mp_to_pauli, QISKIT_OPERATION_MAP
+
+from qiskit_ibm_runtime import QiskitRuntimeService, Session, Estimator
+from qiskit_ibm_runtime.options import Options
+from qiskit_ibm_runtime.constants import RunnerResult
+
+# do not import Estimator (imported above) from qiskit.primitives - the identically
+# named Estimator object has a different call signature than the remote device Estimator,
+# and only runs local simulations. We need the Estimator from qiskit_ibm_runtime. They
+# both use this EstimatorResults, however:
+from qiskit.primitives import EstimatorResult
+
+from qiskit import QuantumCircuit
+
+from qiskit_aer.noise import NoiseModel
+
+
+class Configuration:
+    def __init__(self, n_qubits):
+        self.n_qubits = n_qubits
+        self.noise_model = None
+
+
+class MockedBackend:
+    def __init__(self, num_qubits=10):
+        self._configuration = Configuration(num_qubits)
+        self.options = self._configuration
+        self._service = "SomeServiceProvider"
+        self.name = "mocked_backend"
+
+    def configuration(self):
+        return self._configuration
+
+    def set_options(self, noise_model):
+        self.options.noise_model = noise_model
+
+
+class MockSession:
+    def __init__(self, backend, max_time=None):
+        self.backend = backend
+        self.max_time = max_time
+        self.session_id = "123"
+
+
+try:
+    service = QiskitRuntimeService(channel="ibm_quantum")
+    backend = service.backend("ibmq_qasm_simulator")
+except:
+    backend = MockedBackend()
```

Comment:
is it too messy to make this a pytest fixture? not really sure what the benefit is anyway